File: 187_RepeatedDNASequences.java
//simple but slow
/*
class Solution {
    public List<String> findRepeatedDnaSequences(String s) {
    	HashSet<String> hs = new HashSet<String>();
    	HashSet<String> res = new HashSet<String>();
//        if(s.length() <= 10) return res;
        
        for(int i = 0; i < s.length() - 9; i++) {
        	String cur = s.substring(i, i+10);
        	if(hs.contains(cur)) {
        		res.add(cur);
        	}else {
        		hs.add(cur);
        	}
        }
        return new LinkedList<String>(res);
    }
}
*/

File: c++/187_RepeatedDNASequences.cpp
class Solution {
public:
    vector<string> findRepeatedDnaSequences(string s) {
        unordered_map<string, int> mp;
        for (int i = 0; i + 10 <= s.size(); i++) {
            mp[s.substr(i, 10)]++;
        }
        vector<string> res;
        for (auto [k, v] :  mp) {
            if (v > 1) res.push_back(k);
        }
        return res;
    }
};